File: src/plan_parity.cpp
#include "plan_parity.hpp"

#include <array>
#include <bit>
#include <cmath>
#include <fstream>
#include <limits>
#include <stdexcept>
#include <string>

namespace motionbricks::parity {
namespace {

constexpr std::array<unsigned char, 8> magic{'M','B','P','A','R','I','1',0};
constexpr std::uint32_t version = 1;
constexpr std::uintmax_t maximum_bytes = 512ULL * 1024ULL * 1024ULL;

class reader {
public:
    explicit reader(const std::filesystem::path & path) : stream_(path, std::ios::binary) {
        std::error_code error;
        const auto bytes = std::filesystem::file_size(path, error);
        if (error || bytes < 36U || bytes > maximum_bytes)
            throw std::runtime_error("parity fixture size is outside supported bounds");
        if (!stream_) throw std::runtime_error("could not open parity fixture: " + path.string());
    }
    void exact(void * output, std::size_t bytes, const char * field) {
        stream_.read(static_cast<char *>(output), static_cast<std::streamsize>(bytes));
        if (stream_.gcount() != static_cast<std::streamsize>(bytes))
            throw std::runtime_error(std::string("parity fixture is truncated in ") + field);
    }
    std::uint32_t u32(const char * field) {
        std::array<unsigned char, 4> bytes{}; exact(bytes.data(), bytes.size(), field);
        return std::uint32_t(bytes[0]) | std::uint32_t(bytes[1]) << 8U |
               std::uint32_t(bytes[2]) << 16U | std::uint32_t(bytes[3]) << 24U;
    }
    std::int32_t i32(const char * field) { return std::bit_cast<std::int32_t>(u32(field)); }
    std::uint64_t u64(const char * field) {
        const auto low = u32(field), high = u32(field);
        return low | std::uint64_t(high) << 32U;
    }
    float f32(const char * field) { return std::bit_cast<float>(u32(field)); }
    void eof() {
        char value{};
        if (stream_.get(value)) throw std::runtime_error("parity fixture has trailing bytes");
        if (!stream_.eof()) throw std::runtime_error("could not finish reading parity fixture");
    }
private:
    std::ifstream stream_;
};

std::size_t count(std::initializer_list<std::uint32_t> dimensions, const char * field) {
    std::size_t result = 1;
    for (const auto dimension : dimensions) {
        if (dimension && result > std::numeric_limits<std::size_t>::max() / dimension)
            throw std::runtime_error(std::string("parity dimensions overflow in ") + field);
        result *= dimension;
    }
    return result;
}

template<class T, class Read>
std::vector<T> values(reader & input, std::size_t size, Read method, const char * field) {
    std::vector<T> result; result.reserve(size);
    for (std::size_t index = 0; index < size; ++index) result.push_back((input.*method)(field));
    return result;
}

void finite(const std::vector<float> & data, const char * field) {
    for (const auto value : data)
        if (!std::isfinite(value)) throw std::runtime_error(std::string(field) + " contains a non-finite value");
}

void quaternions(const std::vector<float> & data, const char * field) {
    if (data.size()%4U) throw std::runtime_error(std::string(field)+" has an invalid shape");
    for(std::size_t index=0;index<data.size();index+=4U) {
        const float norm=data[index]*data[index]+data[index+1]*data[index+1]+
                         data[index+2]*data[index+2]+data[index+3]*data[index+3];
        if (!(norm>0.99F && norm<1.01F))
            throw std::runtime_error(std::string(field)+" contains a non-unit quaternion");
    }
}

} // namespace

fixture load(const std::filesystem::path & path) {
    reader input(path);
    std::array<unsigned char, 8> actual{}; input.exact(actual.data(), actual.size(), "magic");
    if (actual != magic || input.u32("version") != version)
        throw std::runtime_error("unsupported parity fixture magic/version");
    fixture result;
    result.fps = input.u32("fps");
    const auto plans = input.u32("plan count");
    const auto joints = input.u32("joint count");
    const auto contexts = input.u32("context frames");
    const auto targets = input.u32("target frames");
    const auto flags = input.u32("flags");
    if (result.fps == 0 || result.fps > 1000 || plans == 0 || plans > 10000 ||
        joints != joint_count || contexts != context_frames || targets != target_frames || flags != 0)
        throw std::runtime_error("parity fixture header dimensions are unsupported");
    result.parents = values<std::int32_t>(input, joint_count, &reader::i32, "parents");
    result.neutral_joints = values<float>(input, joint_count * 3U, &reader::f32, "neutral joints");
    if (result.parents.front() != -1) throw std::runtime_error("parity root parent must be -1");
    for (std::uint32_t joint = 1; joint < joint_count; ++joint)
        if (result.parents[joint] < 0 || result.parents[joint] >= static_cast<std::int32_t>(joint))
            throw std::runtime_error("parity joint topology is invalid");
    result.plans.reserve(plans);
    for (std::uint32_t index = 0; index < plans; ++index) {
        plan item;
        item.command_frame = input.u32("command frame");
        item.expected_frames = input.u32("expected frames");
        item.mode = input.i32("mode");
        const auto reserved = input.u32("reserved");
        item.seed = input.u64("seed");
        for (auto & value : item.movement) value = input.f32("movement");
        for (auto & value : item.facing) value = input.f32("facing");
        if (item.expected_frames < 24 || item.expected_frames > 64 || item.expected_frames % 4U ||
            item.mode < 0 || item.mode >= 15 || reserved != 0)
            throw std::runtime_error("parity plan header is invalid");
        const auto frames = item.expected_frames;
        item.context_roots = values<float>(input, context_frames * 3U, &reader::f32, "context roots");
        item.context_local_rotations = values<float>(
            input, context_frames * joint_count * 4U, &reader::f32, "context rotations");
        item.expected_roots = values<float>(input, count({frames, 3U}, "expected roots"), &reader::f32, "expected roots");
        item.expected_local_rotations = values<float>(
            input, count({frames, joint_count, 4U}, "expected rotations"), &reader::f32, "expected rotations");
        item.expected_joint_positions = values<float>(
            input, count({frames, joint_count, 3U}, "expected joints"), &reader::f32, "expected joints");
        item.target_roots = values<float>(input, target_frames * 3U, &reader::f32, "target roots");
        item.target_local_rotations = values<float>(
            input, target_frames * joint_count * 4U, &reader::f32, "target rotations");
        item.target_joint_positions = values<float>(
            input, target_frames * joint_count * 3U, &reader::f32, "target joints");
        finite(item.context_roots, "context roots"); finite(item.context_local_rotations, "context rotations");
        finite(item.expected_roots, "expected roots"); finite(item.expected_local_rotations, "expected rotations");
        finite(item.expected_joint_positions, "expected joints"); finite(item.target_roots, "target roots");
        finite(item.target_local_rotations, "target rotations"); finite(item.target_joint_positions, "target joints");
        quaternions(item.context_local_rotations,"context rotations");
        quaternions(item.expected_local_rotations,"expected rotations");
        quaternions(item.target_local_rotations,"target rotations");
        result.plans.push_back(std::move(item));
    }
    finite(result.neutral_joints, "neutral joints");
    input.eof();
    return result;
}

} // namespace motionbricks::parity

File: reference/plan_parity.py
#!/usr/bin/env python3
"""Bounded binary format for independent open-loop agent parity plans."""

from __future__ import annotations

import struct
from dataclasses import dataclass
from pathlib import Path

import numpy as np


MAGIC = b"MBPARI1\0"
VERSION = 1
FPS = 30
JOINTS = 34
CONTEXT_FRAMES = 4
TARGET_FRAMES = 4
HEADER = struct.Struct("<8s7I")
PLAN_HEADER = struct.Struct("<IIiIQ6f")


@dataclass
class Plan:
    command_frame: int
    expected_frames: int
    mode: int
    seed: int
    movement: np.ndarray
    facing: np.ndarray
    context_roots: np.ndarray
    context_local_rotations: np.ndarray
    expected_roots: np.ndarray
    expected_local_rotations: np.ndarray
    expected_joint_positions: np.ndarray
    target_roots: np.ndarray
    target_local_rotations: np.ndarray
    target_joint_positions: np.ndarray


@dataclass
class Fixture:
    parents: np.ndarray
    neutral_joints: np.ndarray
    plans: list[Plan]
    fps: int = FPS


def _array(value, dtype, shape: tuple[int, ...], name: str) -> np.ndarray:
    result = np.asarray(value, dtype=dtype)
    if result.shape != shape:
        raise ValueError(f"{name} has shape {result.shape}, expected {shape}")
    if np.issubdtype(result.dtype, np.floating) and not np.isfinite(result).all():
        raise ValueError(f"{name} contains a non-finite value")
    return np.ascontiguousarray(result)


def _quaternions(value: np.ndarray, name: str) -> None:
    norms = np.sum(value * value, axis=-1)
    if np.any(norms <= 0.99) or np.any(norms >= 1.01):
        raise ValueError(f"{name} contains a non-unit quaternion")


def validate(fixture: Fixture) -> Fixture:
    if fixture.fps <= 0 or fixture.fps > 1000 or not fixture.plans or len(fixture.plans) > 10000:
        raise ValueError("fixture dimensions are outside supported bounds")
    fixture.parents = _array(fixture.parents, "<i4", (JOINTS,), "parents")
    fixture.neutral_joints = _array(
        fixture.neutral_joints, "<f4", (JOINTS, 3), "neutral_joints"
    )
    if fixture.parents[0] != -1:
        raise ValueError("root parent must be -1")
    for joint, parent in enumerate(fixture.parents):
        if joint and (parent < 0 or parent >= joint):
            raise ValueError(f"invalid parent {parent} for joint {joint}")
    for index, plan in enumerate(fixture.plans):
        prefix = f"plan[{index}]"
        if not (24 <= plan.expected_frames <= 64 and plan.expected_frames % 4 == 0):
            raise ValueError(f"{prefix} has an invalid expected frame count")
        if plan.mode < 0 or plan.mode >= 15 or plan.seed < 0 or plan.seed >= 1 << 64:
            raise ValueError(f"{prefix} has invalid command metadata")
        plan.movement = _array(plan.movement, "<f4", (3,), prefix + ".movement")
        plan.facing = _array(plan.facing, "<f4", (3,), prefix + ".facing")
        plan.context_roots = _array(
            plan.context_roots, "<f4", (CONTEXT_FRAMES, 3), prefix + ".context_roots"
        )
        plan.context_local_rotations = _array(
            plan.context_local_rotations, "<f4", (CONTEXT_FRAMES, JOINTS, 4),
            prefix + ".context_local_rotations",
        )
        _quaternions(plan.context_local_rotations, prefix + ".context_local_rotations")
        frames = plan.expected_frames
        plan.expected_roots = _array(
            plan.expected_roots, "<f4", (frames, 3), prefix + ".expected_roots"
        )
        plan.expected_local_rotations = _array(
            plan.expected_local_rotations, "<f4", (frames, JOINTS, 4),
            prefix + ".expected_local_rotations",
        )
        _quaternions(plan.expected_local_rotations, prefix + ".expected_local_rotations")
        plan.expected_joint_positions = _array(
            plan.expected_joint_positions, "<f4", (frames, JOINTS, 3),
            prefix + ".expected_joint_positions",
        )
        plan.target_roots = _array(
            plan.target_roots, "<f4", (TARGET_FRAMES, 3), prefix + ".target_roots"
        )
        plan.target_local_rotations = _array(
            plan.target_local_rotations, "<f4", (TARGET_FRAMES, JOINTS, 4),
            prefix + ".target_local_rotations",
        )
        _quaternions(plan.target_local_rotations, prefix + ".target_local_rotations")
        plan.target_joint_positions = _array(
            plan.target_joint_positions, "<f4", (TARGET_FRAMES, JOINTS, 3),
            prefix + ".target_joint_positions",
        )
    return fixture


def write_fixture(path: Path, fixture: Fixture) -> None:
    fixture = validate(fixture)
    with path.open("xb") as stream:
        stream.write(HEADER.pack(
            MAGIC, VERSION, fixture.fps, len(fixture.plans), JOINTS,
            CONTEXT_FRAMES, TARGET_FRAMES, 0,
        ))
        stream.write(fixture.parents.tobytes())
        stream.write(fixture.neutral_joints.tobytes())
        for plan in fixture.plans:
            stream.write(PLAN_HEADER.pack(
                plan.command_frame, plan.expected_frames, plan.mode, 0, plan.seed,
                *plan.movement.tolist(), *plan.facing.tolist(),
            ))
            for value in (
                plan.context_roots, plan.context_local_rotations,
                plan.expected_roots, plan.expected_local_rotations,
                plan.expected_joint_positions, plan.target_roots,
                plan.target_local_rotations, plan.target_joint_positions,
            ):
                stream.write(value.tobytes(order="C"))


def read_fixture(path: Path) -> Fixture:
    raw = path.read_bytes()
    if len(raw) < HEADER.size:
        raise ValueError("fixture is shorter than its header")
    magic, version, fps, count, joints, context, targets, flags = HEADER.unpack_from(raw)
    if (magic != MAGIC or version != VERSION or joints != JOINTS or
            context != CONTEXT_FRAMES or targets != TARGET_FRAMES or flags != 0):
        raise ValueError("unsupported parity fixture header")
    if fps <= 0 or fps > 1000 or count <= 0 or count > 10000:
        raise ValueError("fixture header dimensions are outside supported bounds")
    offset = HEADER.size

    def take(dtype, shape: tuple[int, ...], name: str) -> np.ndarray:
        nonlocal offset
        size = int(np.prod(shape)) * np.dtype(dtype).itemsize
        if size > len(raw) - offset:
            raise ValueError(f"fixture is truncated in {name}")
        result = np.frombuffer(raw, dtype=dtype, count=int(np.prod(shape)), offset=offset)
        offset += size
        return result.reshape(shape).copy()

    parents = take("<i4", (JOINTS,), "parents")
    neutral = take("<f4", (JOINTS, 3), "neutral joints")
    plans: list[Plan] = []
    for index in range(count):
        if PLAN_HEADER.size > len(raw) - offset:
            raise ValueError(f"fixture is truncated in plan {index} header")
        values = PLAN_HEADER.unpack_from(raw, offset)
        offset += PLAN_HEADER.size
        command_frame, frames, mode, reserved, seed, *directions = values
        if reserved != 0 or frames < 24 or frames > 64 or frames % 4:
            raise ValueError(f"fixture has an invalid plan {index} header")
        plans.append(Plan(
            command_frame=command_frame, expected_frames=frames, mode=mode, seed=seed,
            movement=np.asarray(directions[:3], dtype=np.float32),
            facing=np.asarray(directions[3:], dtype=np.float32),
            context_roots=take("<f4", (CONTEXT_FRAMES, 3), "context roots"),
            context_local_rotations=take(
                "<f4", (CONTEXT_FRAMES, JOINTS, 4), "context rotations"
            ),
            expected_roots=take("<f4", (frames, 3), "expected roots"),
            expected_local_rotations=take(
                "<f4", (frames, JOINTS, 4), "expected rotations"
            ),
            expected_joint_positions=take(
                "<f4", (frames, JOINTS, 3), "expected positions"
            ),
            target_roots=take("<f4", (TARGET_FRAMES, 3), "target roots"),
            target_local_rotations=take(
                "<f4", (TARGET_FRAMES, JOINTS, 4), "target rotations"
            ),
            target_joint_positions=take(
                "<f4", (TARGET_FRAMES, JOINTS, 3), "target positions"
            ),
        ))
    if offset != len(raw):
        raise ValueError(f"fixture has {len(raw) - offset} trailing bytes")
    return validate(Fixture(parents=parents, neutral_joints=neutral, plans=plans, fps=fps))
